include/nbl/builtin/hlsl/morton.hlsl

```diff
@@ -0,0 +1,698 @@
+
+template<uint16_t Dim, uint16_t Bits, typename encode_t>
+struct MortonEncoder;
+
+template<uint16_t Bits, typename encode_t>
+struct MortonEncoder<2, Bits, encode_t>
+{
+    template<typename decode_t>
+    NBL_CONSTEXPR_STATIC_INLINE_FUNC encode_t encode(NBL_CONST_REF_ARG(decode_t) decodedValue)
+    {
+        left_shift_operator<encode_t> leftShift;
+        encode_t encoded = _static_cast<encode_t>(decodedValue);
+        NBL_IF_CONSTEXPR(Bits > 16)
+        {
+            encoded = (encoded | leftShift(encoded, 16)) & morton_mask_2_4_v<encode_t>;
+        }
+        NBL_IF_CONSTEXPR(Bits > 8)
+        {
+            encoded = (encoded | leftShift(encoded, 8)) & morton_mask_2_3_v<encode_t>;
+        }
+        NBL_IF_CONSTEXPR(Bits > 4)
+        {
+            encoded = (encoded | leftShift(encoded, 4)) & morton_mask_2_2_v<encode_t>;
+        }
+        NBL_IF_CONSTEXPR(Bits > 2)
+        {
+            encoded = (encoded | leftShift(encoded, 2)) & morton_mask_2_1_v<encode_t>;
+        }
+        encoded = (encoded | leftShift(encoded, 1)) & morton_mask_2_0_v<encode_t>;
+        return encoded;
+    }
+};
+
+template<uint16_t Bits, typename encode_t>
+struct MortonEncoder<3, Bits, encode_t>
+{
+    template<typename decode_t>
+    NBL_CONSTEXPR_STATIC_INLINE_FUNC encode_t encode(NBL_CONST_REF_ARG(decode_t) decodedValue)
+    {
+        left_shift_operator<encode_t> leftShift;
+        encode_t encoded = _static_cast<encode_t>(decodedValue);
+        NBL_IF_CONSTEXPR(Bits > 12)
+        {
+            encoded = (encoded | leftShift(encoded, 24)) & morton_mask_3_3_v<encode_t>;
+        }
+        NBL_IF_CONSTEXPR(Bits > 6)
+        {
+            encoded = (encoded | leftShift(encoded, 12)) & morton_mask_3_2_v<encode_t>;
+        }
+        NBL_IF_CONSTEXPR(Bits > 3)
+        {
+            encoded = (encoded | leftShift(encoded, 6)) & morton_mask_3_1_v<encode_t>;
+        }
+        encoded = (encoded | leftShift(encoded, 2) | leftShift(encoded, 4)) & morton_mask_3_0_v<encode_t>;
+        return encoded;
+    }
+};
+
+template<uint16_t Bits, typename encode_t>
+struct MortonEncoder<4, Bits, encode_t>
+{
+    template<typename decode_t>
+    NBL_CONSTEXPR_STATIC_INLINE_FUNC encode_t encode(NBL_CONST_REF_ARG(decode_t) decodedValue)
+    {
+        left_shift_operator<encode_t> leftShift;
+        encode_t encoded = _static_cast<encode_t>(decodedValue);
+        NBL_IF_CONSTEXPR(Bits > 8)
+        {
+            encoded = (encoded | leftShift(encoded, 24)) & morton_mask_4_3_v<encode_t>;
+        }
+        NBL_IF_CONSTEXPR(Bits > 4)
+        {
+            encoded = (encoded | leftShift(encoded, 12)) & morton_mask_4_2_v<encode_t>;
+        }
+        NBL_IF_CONSTEXPR(Bits > 2)
+        {
+            encoded = (encoded | leftShift(encoded, 6)) & morton_mask_4_1_v<encode_t>;
+        }
+        encoded = (encoded | leftShift(encoded, 3)) & morton_mask_4_0_v<encode_t>;
```

Comment:
I think if you heed https://github.com/Devsh-Graphics-Programming/Nabla/pull/860/files/e25a35cce8f0554baf98173f9cc1d1dd93629042..60ff99a4dadfdecc5abf59e4fb2d95e62d6ed929#r2032928204 then you can avoid having explicit specializations for every dimension

Reply:
also check the optimized SPIR-V if these if statements eliminate with `-O1` optimization level

Reply:
also check `Bits>1` see https://github.com/Devsh-Graphics-Programming/Nabla/pull/860/files/e25a35cce8f0554baf98173f9cc1d1dd93629042..60ff99a4dadfdecc5abf59e4fb2d95e62d6ed929#r2033023021